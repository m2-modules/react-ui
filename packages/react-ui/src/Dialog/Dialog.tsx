import { createPortal } from "react-dom";
import Overlay, { type OverlayProps } from "../Overlay";
import { useCallback, useEffect, useRef, useState } from "react";
import usePortal from "../hooks/usePortal";

export interface DialogProps {
  active?: boolean;
  enableOverlay?: boolean;
  activateTransitionClasses?: string[];
  inactivateTransitionClasses?: string[];
  transitionDuration?: string;
  onClose: () => void;
  overlayProps?: Omit<OverlayProps, "active" | "onClose">;
  children: React.ReactNode;
}

export default function Dialog({
  active = false,
  enableOverlay = true,
  activateTransitionClasses = ["opacity-1"],
  inactivateTransitionClasses = ["opacity-0"],
  transitionDuration = "duration-500",
  onClose,
  overlayProps,
  children,
}: DialogProps) {
  const dialogContainerRef = useRef<HTMLDivElement>(null);
  const dialogPortal = usePortal("dialog-portal");
  const [_active, _setActive] = useState(active);

  const activateDialog = useCallback(() => {
    if (!dialogContainerRef.current) return;
    const dialogContainer = dialogContainerRef.current;
    _setActive(true);

    setTimeout(() => {
      dialogContainer.classList.remove(...inactivateTransitionClasses);
      dialogContainer.classList.add(...activateTransitionClasses);
    }, 100);
  }, [activateTransitionClasses, inactivateTransitionClasses]);

  const inactivateDialog = useCallback(() => {
    if (!dialogContainerRef.current) return;
    const dialogContainer = dialogContainerRef.current;

    dialogContainer.classList.remove(...activateTransitionClasses);
    dialogContainer.classList.add(...inactivateTransitionClasses);
  }, [activateTransitionClasses, inactivateTransitionClasses]);

  useEffect(() => {
    if (active) {
      activateDialog();
    } else {
      inactivateDialog();
    }
  }, [activateDialog, active, inactivateDialog]);

  if ((!active && !_active) || !dialogPortal) return <></>;

  return createPortal(
    <>
      {enableOverlay && (
        <Overlay active={active} onClose={onClose} {...overlayProps} />
      )}
      <div className="fixed left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2 z-[3]">
        <div
          ref={dialogContainerRef}
          className={`transition-all ${transitionDuration} ${inactivateTransitionClasses.join(
            " ",
          )}`}
          onTransitionEnd={() => {
            if (!active && _active) _setActive(false);
          }}
        >
          {children}
        </div>
      </div>
    </>,
    dialogPortal,
  );
}
