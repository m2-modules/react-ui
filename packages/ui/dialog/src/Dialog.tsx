import { createPortal } from "react-dom";
import Overlay from "@m2-modules/overlay";
import { useCallback, useEffect, useRef, useState } from "react";

const ACTIVATE_CLASS_NAME = "opacity-1";
const INACTIVATE_CLASS_NAME = "opacity-0";

const getDialogPortal = () => {
  let dialogPortal = document.querySelector<HTMLDivElement>("#dialog-portal");
  if (dialogPortal) return dialogPortal;
  dialogPortal = document.createElement("div");
  dialogPortal.id = "dialog-portal";
  document.body.append(dialogPortal);
  return dialogPortal;
};

export interface DialogProps {
  active?: boolean;
  enableOverlay?: boolean;
  onClose: () => void;
  children: React.ReactNode;
}

export default function Dialog({
  active = false,
  enableOverlay = true,
  onClose,
  children,
}: DialogProps) {
  const dialogContainerRef = useRef<HTMLDivElement>(null);
  const [_active, _setActive] = useState(active);

  const activateDialog = useCallback(() => {
    if (!dialogContainerRef.current) return;
    const dialogContainer = dialogContainerRef.current;
    _setActive(true);

    setTimeout(() => {
      dialogContainer.classList.remove(INACTIVATE_CLASS_NAME);
      dialogContainer.classList.add(ACTIVATE_CLASS_NAME);
    }, 100);
  }, []);

  const inactivateDialog = useCallback(() => {
    if (!dialogContainerRef.current) return;
    const dialogContainer = dialogContainerRef.current;

    dialogContainer.classList.remove(ACTIVATE_CLASS_NAME);
    dialogContainer.classList.add(INACTIVATE_CLASS_NAME);
  }, []);

  useEffect(() => {
    if (active) {
      activateDialog();
    } else {
      inactivateDialog();
    }
  }, [activateDialog, active, inactivateDialog]);

  if (!active && !_active) return <></>;

  return createPortal(
    <>
      {enableOverlay && <Overlay active={active} onClose={onClose} />}
      <div
        ref={dialogContainerRef}
        className="fixed left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2 transition-opacity duration-500 opacity-0 z-[3]"
        onTransitionEnd={() => {
          if (!active && _active) _setActive(false);
        }}
      >
        {children}
      </div>
    </>,
    getDialogPortal(),
  );
}
