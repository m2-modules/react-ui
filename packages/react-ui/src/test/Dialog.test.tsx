import renderer from "react-test-renderer";
import Dialog from "../Dialog";

it("show Dialog on document", () => {
  const component = renderer.create(
    <Dialog
      onClose={() => {
        console.log("on close");
      }}
    >
      <h1>Dialog</h1>
    </Dialog>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
