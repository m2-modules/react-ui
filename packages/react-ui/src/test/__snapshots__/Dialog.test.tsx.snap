// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show Dialog on document 1`] = `null`;
